### ISSUE TITLE
Better control over which window files are opened into

### ISSUE THREAD

**Issue by jpmhouston** [zed-industries/zed#18105]
### Check for existing issues

- [X] Completed

### Describe the feature

I would like some control over which windows files are opened into. For example, I usually have a non-project window open with a couple files I like to be able to quickly get to. With a window like this open, every file I open from the Open menu item, from opening by the OS (I'm macOS so I mean the Finder), or from the command line helper, they all open as tabs in this same window. 

I would like to be able to configure zed to get this different behavior:
- files I open from outside of the app to open in a fresh window (and on macOS for this window to open into the current space, as long as the app isn't assigned in the dock to a particular other space I guess)
- opening multiple files at once all open to tabs in the same new window

Ideally there should be some way to explcitly open files into the current window instead a new one. I would be happy for my use case if, when zed is active, using Open from the menu opened into a tab to the frontmost window as I could always open a new window and then use Open to get the alternate result.

**Comment by notpeter** [comment]
I believe you can do some of this from the `zed` CLI today. Specifically:
- `zed <directory>` will open a project with that directory
  - If there is an existing open Zed window for that that directory it will just bring that window to the foreground
  - `zed -n <directory>` will open a new workspace and won't attempt to reuse existing workspaces
- `zed <directory> <filename>` Will do the same as above, but it will also open the filename specified (which need not already exist).

Alternatively, if you have a finder window open you can select one or multiple files, drag them into the Zed Icon in the dock (wait a second) and then drag them into individual open workspaces.  If you need a new empty Workspace, you can always go File->New Window (`cmd-shift-n`) before you drag or `cmd-o`.

As for the MacOS Spaces-specific behavior, I personally don't use spaces. If you have specific actionable suggestions to make Zed behave more like other specific apps that handle Spaces well, I'm all ears.

---

If I understand you correctly you want Zed when launched from the Finder or cli 

1. when opening multiple files, always spawn a new workspace without a project folder
2. when opening a file, reuse an existing workspace if there is a workspace with a project folder that is a parent of this file, otherwise:
3. when opening a file, spawn a new workspace without a project folder

Do I have that correct?

Basically, I want to make all sure any enhancement issue are scoped and actionable.
If you'd prefer I can move this to a discussion to explore options.

**Comment by jpmhouston** [comment]
Yes, 1 and 3 captures my request well, and yes 2 makes sense to include as part of my desired workflow.

**Comment by Petemir** [comment]
In the original issue, it says

> I would be happy for my use case if, when zed is active, using Open from the menu opened into a tab to the frontmost window as I could always open a new window and then use Open to get the alternate result.

I think that request (i.e., opening from a current Zed window) would be out of the following actionable scope, which deals open calls with Finder/cli:

> If I understand you correctly you want Zed when launched from the Finder or cli
> 
>     1. when opening multiple files, always spawn a new workspace without a project folder
> 
>     2. when opening a file, reuse an existing workspace if there is a workspace with a project folder that is a parent of this file, otherwise:
> 
>     3. when opening a file, spawn a new workspace without a project folder

My issue: I have two open windows with two unrelated projects. I had to exclude a file from one of those projects for performance reasons (#18300), but I still want to open it, so I then use the open dialog (not Finder nor cli). That file always ends up in the unrelated window. 

2. _could_ fit as a solution, but if it is restricted to only Finder/cli then it wouldn't.

**Comment by scottyeager** [comment]
> 3\. when opening a file, spawn a new workspace without a project folder

The behavior I want in this case is for the file to open in the last Zed window that had active focus. This is how my web browser works, for example—if I open a link or file from the cli, it opens in the last window I visited.

Current behavior in Zed seems to be that files not belonging to any currently opened workspace always get opened in the same window (the oldest window?). This is basically as good as random for whether it matches user intent and expectation for where the new tab will show up.

**Comment by AshkanArabim** [comment]
Having the option to open independent files in their own window would AMAZING for me, the GNOME user.

Right now, if I have Zed open in another workspace, and decide to double click a file to open it in zed, it kicks me to the workspace where zed is open and opens the new file as a tab. This sucks very much. I have to detach the tab and move it to where I was previously working. 

In VSCode there's a setting called `window.openFilesInNewWindow` which solves this exact issue. It would be nice to have the equivalent on Zed.

**Comment by dflock** [comment]
Yeah, some kind of "quikcly edit tyhis one-off file" mode is what I want. Just making the `-n` cli option work for files would do it. If I just want to edit some random config file real quick, I can then just say `zed -n file.cfg` - and get a new Zed window, right here on the current minoter & workspace, with that file in it - no projects, no switching to the last active zed, no switching workspaces.

**Comment by JosephTLyons** [comment]
Hello,

We’re closing this issue as part of a cleanup of older feature requests.

Going forward, new feature requests start as GitHub Discussions and only become Issues once the direction is agreed on. This helps ensure that Issues represent scoped, ready-to-build work and makes it easier for contributors to find work that’s ready to pick up. You can read more in our [Feature Request Guidelines](https://github.com/zed-industries/zed/discussions/47963).

As part of this transition, we’re closing feature request Issues that are older than 6 months and have fewer than 10 upvotes.

If this feature hasn’t been implemented and is still important to you, we encourage you to open a new feature request Discussion [here](https://github.com/zed-industries/zed/discussions/new?category=feature-requests). That’s where we track and evaluate new requests.

Thanks for helping shape the future of Zed.
